### ISSUE TITLE
buildbot stop is not reliable in buildbot 0.9

### ISSUE THREAD

**Issue by aj062** [buildbot/buildbot#3535]
If have noticed multiple times that buildbot stop never finish. Running 'buildbot stop' again results in Error "twisted.internet.error.ReactorNotRunning: Can't stop reactor that isn't running". Only option left at that time is to "kill -9 pid".

'buildbot stop' should be more reliable.

For example, below are logs from a test instance (single master) which was not running any build at the time of 'buildbot stop'. However, 'buildbot stop' never finished, issuing another 'buildbot stop' didn't help either.

```
2017-08-16 09:42:23-0700 [-] Received SIGTERM, shutting down.
2017-08-16 09:42:23-0700 [-] doing housekeeping for master 1 build.hostname.com:[path_here]
2017-08-16 09:42:24-0700 [-] Initiating clean shutdown
2017-08-16 09:42:24-0700 [-] No running jobs, starting shutdown immediately
2017-08-16 09:42:24-0700 [-] (ignored) while invoking Triggerable schedulers:
	Traceback (most recent call last):
	Failure: exceptions.RuntimeError: Triggerable scheduler stopped before build was complete
	
2017-08-16 09:42:24-0700 [-] (ignored) while invoking Triggerable schedulers:
	Traceback (most recent call last):
	Failure: exceptions.RuntimeError: Triggerable scheduler stopped before build was complete
	
2017-08-16 09:42:24-0700 [-] (ignored) while invoking Triggerable schedulers:
	Traceback (most recent call last):
	Failure: exceptions.RuntimeError: Triggerable scheduler stopped before build was complete
	
2017-08-16 09:42:24-0700 [-] (ignored) while invoking Triggerable schedulers:
	Traceback (most recent call last):
	Failure: exceptions.RuntimeError: Triggerable scheduler stopped before build was complete
	
2017-08-16 09:42:24-0700 [-] (ignored) while invoking Triggerable schedulers:
	Traceback (most recent call last):
	Failure: exceptions.RuntimeError: Triggerable scheduler stopped before build was complete
	
2017-08-16 09:42:24-0700 [-] (ignored) while invoking Triggerable schedulers:
	Traceback (most recent call last):
	Failure: exceptions.RuntimeError: Triggerable scheduler stopped before build was complete
	
2017-08-16 09:42:24-0700 [-] (ignored) while invoking Triggerable schedulers:
	Traceback (most recent call last):
	Failure: exceptions.RuntimeError: Triggerable scheduler stopped before build was complete
	
2017-08-16 09:42:24-0700 [-] (ignored) while invoking Triggerable schedulers:
	Traceback (most recent call last):
	Failure: exceptions.RuntimeError: Triggerable scheduler stopped before build was complete
	
2017-08-16 09:42:25-0700 [-] (ignored) while invoking Triggerable schedulers:
	Traceback (most recent call last):
	Failure: exceptions.RuntimeError: Triggerable scheduler stopped before build was complete
	
2017-08-16 09:42:25-0700 [-] (ignored) while invoking Triggerable schedulers:
	Traceback (most recent call last):
	Failure: exceptions.RuntimeError: Triggerable scheduler stopped before build was complete
	
2017-08-16 09:42:25-0700 [-] (ignored) while invoking Triggerable schedulers:
	Traceback (most recent call last):
	Failure: exceptions.RuntimeError: Triggerable scheduler stopped before build was complete
	
2017-08-16 09:42:25-0700 [-] (ignored) while invoking Triggerable schedulers:
	Traceback (most recent call last):
	Failure: exceptions.RuntimeError: Triggerable scheduler stopped before build was complete
	
2017-08-16 09:42:25-0700 [-] (ignored) while invoking Triggerable schedulers:
	Traceback (most recent call last):
	Failure: exceptions.RuntimeError: Triggerable scheduler stopped before build was complete
	
2017-08-16 09:42:25-0700 [-] (ignored) while invoking Triggerable schedulers:
	Traceback (most recent call last):
	Failure: exceptions.RuntimeError: Triggerable scheduler stopped before build was complete
	
2017-08-16 09:42:25-0700 [-] (ignored) while invoking Triggerable schedulers:
	Traceback (most recent call last):
	Failure: exceptions.RuntimeError: Triggerable scheduler stopped before build was complete
	
2017-08-16 09:42:25-0700 [-] (ignored) while invoking Triggerable schedulers:
	Traceback (most recent call last):
	Failure: exceptions.RuntimeError: Triggerable scheduler stopped before build was complete
	
2017-08-16 09:42:25-0700 [-] (ignored) while invoking Triggerable schedulers:
	Traceback (most recent call last):
	Failure: exceptions.RuntimeError: Triggerable scheduler stopped before build was complete
	
2017-08-16 09:42:25-0700 [-] (ignored) while invoking Triggerable schedulers:
	Traceback (most recent call last):
	Failure: exceptions.RuntimeError: Triggerable scheduler stopped before build was complete
	
2017-08-16 09:42:25-0700 [-] (ignored) while invoking Triggerable schedulers:
	Traceback (most recent call last):
	Failure: exceptions.RuntimeError: Triggerable scheduler stopped before build was complete
	
2017-08-16 09:42:25-0700 [-] (ignored) while invoking Triggerable schedulers:
	Traceback (most recent call last):
	Failure: exceptions.RuntimeError: Triggerable scheduler stopped before build was complete
	
2017-08-16 09:42:25-0700 [-] (ignored) while invoking Triggerable schedulers:
	Traceback (most recent call last):
	Failure: exceptions.RuntimeError: Triggerable scheduler stopped before build was complete
	
2017-08-16 09:42:25-0700 [-] (ignored) while invoking Triggerable schedulers:
	Traceback (most recent call last):
	Failure: exceptions.RuntimeError: Triggerable scheduler stopped before build was complete
	
2017-08-16 09:42:26-0700 [-] (TCP Port 8010 Closed)
2017-08-16 09:42:26-0700 [-] Stopping factory <buildbot.www.service.BuildbotSite instance at 0x7ff4580b0dd0>
2017-08-16 09:42:26-0700 [-] (TCP Port 9989 Closed)
2017-08-16 09:42:26-0700 [-] Stopping factory <twisted.spread.pb.PBServerFactory instance at 0x3545878>



2017-08-16 09:51:21-0700 [-] Received SIGTERM, shutting down.
2017-08-16 09:51:21-0700 [-] Unhandled Error
	Traceback (most recent call last):
	  File "/usr/lib64/python2.7/site-packages/twisted/application/app.py", line 396, in startReactor
	    self.config, oldstdout, oldstderr, self.profiler, reactor)
	  File "/usr/lib64/python2.7/site-packages/twisted/application/app.py", line 311, in runReactorWithLogging
	    reactor.run()
	  File "/usr/lib64/python2.7/site-packages/twisted/internet/base.py", line 1243, in run
	    self.mainLoop()
	  File "/usr/lib64/python2.7/site-packages/twisted/internet/base.py", line 1252, in mainLoop
	    self.runUntilCurrent()
	--- <exception caught here> ---
	  File "/usr/lib64/python2.7/site-packages/twisted/internet/base.py", line 851, in runUntilCurrent
	    f(*a, **kw)
	  File "/usr/lib64/python2.7/site-packages/twisted/internet/base.py", line 627, in stop
	    "Can't stop reactor that isn't running.")
	twisted.internet.error.ReactorNotRunning: Can't stop reactor that isn't running.
```

**Comment by tardyp** [comment]
This is a known issue related to txrequests, and cleanup of txrequests threads. I was never able to find a reliable method to shutdown the threads correctly sometimes it works some times it doesn't..

This is why I added support to treq, which doesn't use threads. You may switch to treq as a backend for httpclientservice.
```
pip install treq
pip uninstall txrequests
```

**Comment by aj062** [comment]
We seems to be using treq. (Though I never manually installed any of them. Probably buildbot by default used treq in my case).

[~]$pip freeze | grep txrequests
[~]$
[~]$pip freeze | grep treq
treq==17.7.0

Do you expect this problem with treq as well?

**Comment by seankelly** [comment]
Are you using any latent workers? EC2 and OpenStack use threads I recall.

**Comment by aj062** [comment]
@seankelly no.

@tardyp This problem happens frequently on heavily loaded master. I think the problem is that even after running `buildbot stop` command, new builds keeps starting. So the load doesn't decrease and there is always long list of pending_builders, and so `buildbot stop` is not able to acquire activity_lock.

The **workaround** which works for me is:
`buildbot stop --clean`
`#wait for 3-4 minutes and then press control+c`
`buildbot stop`

`buildbot stop --clean` helps in reducing the load, and then `buildbot stop` works.

I think new builds should not start once `buildbot stop` is issued.

**Comment by tardyp** [comment]
> The workaround which works for me is
> [..]
ahah. Interesting findings. Do you still have my change in  the brd in your prod?
https://github.com/buildbot/buildbot/pull/3684/files

The activity lock should be taken much less longer there.

**Comment by aj062** [comment]
@tardyp The workaround I mentioned above (using --clean) has been working perfectly fine for me for last few months. I just tried to do regular "buildbot stop" and it failed again (as expected).

> ahah. Interesting findings. Do you still have my change in the brd in your prod?

Yes, I have that change. 

Also, I am not not sure if buildbot is waiting for activity_lock or stuck somewhere else. I notice the line "doing housekeeping for master" in the logs. I have added a log statement at https://git.io/vdKlC, but that wasn't printed in the logs (atleast this time). So, buildbot might be stuck in _masterDeactivatedHousekeeping() https://git.io/vAWTf or somewhere else. Note that the number of builds running currently doesn't decrease for me (except for few builds which finish normally).

Here are the redacted twistd.log: https://goo.gl/NwzJcC

I think that one problem which is surely present is that even after running buildbot stop command, new builds keeps starting. "buildbot stop" behavior should be same as "buildbot stop --clean", in the sense that new builds shouldn't start after issuing stop command.

**Comment by aj062** [comment]
@tardyp, for sure, "buildbot stop --clean" prevents new builds from starting, while "buildbot stop" doesn't. Can you please check why?

**Comment by aj062** [comment]
@tardyp any idea why "buildbot stop" doesn't prevents new builds from starting, while "buildbot stop --clean" does?

**Comment by aj062** [comment]
Make sense.

Is this the code which prevents new builds from starting (while using "buildbot stop --clean")?

https://github.com/buildbot/buildbot/blob/master/master/buildbot/process/botmaster.py#L105

**Comment by seankelly** [comment]
That highlighted line adds all running builds to a list so they can be monitored. I don't know where builds are queued, it appears to be elsewhere than this file. My guess is the BuildRequestDistributor can't distribute them because it's disowned so the builds naturally queue.

**Comment by openSourceBugs** [comment]
Wow.  5 year old issue and this is still happening for me.  I probably can't use buildbot at this point.  I'm not even on section 1.2.3 of the tutorial and open source quality rears its head.

**Comment by p12tic** [comment]
@openSourceBugs Which version are you using. The reliability of `buildbot stop` has been quite improved around version 2.10.2 / 3.0 released 1.5 years ago.

**Comment by openSourceBugs** [comment]
> @openSourceBugs Which version are you using? The reliability of `buildbot stop` has been quite improved around version 2.10.2 / 3.0 released 1.5 years ago.
> 
> Note that you need to run `buildbot stop` with `--no-wait` in order for it to not wait for builds to complete.

This was on what I pip installed yesterday from the tutorial docs.  There was a "waiting for build to stop" status that never went away for 30 minutes after attempting to stop a failed build.  Me stopping the build arose because the tutorial and the code it is using are obsolete (using git:// instead of https) and hasn't been updated for 5 years at least.  There was a timeout when running the tutorial as it is now.  I probably won't be using buildbot moving forward and this will be my last comment on this.

**Comment by aj062** [comment]
> The reliability of buildbot stop has been quite improved around version 2.10.2 / 3.0 released 1.5 years ago.

Interesting. It would be nice if https://github.com/buildbot/buildbot/issues/6029 can also be fixed.

**Comment by mokibit** [comment]
For the record, the issue with using obsolete code in tutorial was recently fixed: https://github.com/buildbot/buildbot/pull/6599.

**Comment by mokibit** [comment]
This issue was fixed in #6620.
